lib/ptracer.h

```diff
@@ -0,0 +1,471 @@
+   get called. Therefore, the stack will get corrupted if the previous
+   call-frames are not removed. The finalizer function makes sure that
+   it does not happen. It's guardian angel. */
+/* The finalizer function will be called from a to-be-closed value (since
+   Lua 5.4). If you are using Lua version prior 5.4, you are outta luck. */
+int pallene_tracer_finalizer(lua_State *L) {
+    /* Get the userdata. */
+    pt_cont_t *cont = (pt_cont_t *) lua_touserdata(L, lua_upvalueindex(1));
+
+    /* Remove all the frames until last Lua frame. */
+    int idx = cont->count - 1;
+    while(cont->stack[idx].type != PALLENE_TRACER_FRAME_TYPE_LUA)
+        idx--;
+
+    /* Remove the Lua frame as well. */
+    cont->count = idx;
+
+    return 0;
+}
+
+/* Initializes the Pallene Tracer. The initialization refers to creating the stack
+   if not created, preparing the traceback fn and finalizers. */
+/* NOTE: Pushes the finalizer object to the stack. The object has to be closed
+   everytime you are in a Lua C funciton using `lua_toclose(L, idx)` function. */
+pt_cont_t *pallene_tracer_init(lua_State *L) {
+    pt_cont_t *cont = NULL;
+
+    /* Try getting the userdata. */
+    lua_getfield(L, LUA_REGISTRYINDEX, PALLENE_TRACER_CONTAINER_ENTRY);
+
+    /* If we don't find any userdata, initialize resources. */
+    if(luai_unlikely(lua_isnil(L, -1) == 1)) {
+        cont = (pt_cont_t *) lua_newuserdata(L, sizeof(pt_cont_t));
+        cont->stack = malloc(PALLENE_TRACER_MAX_CALLSTACK * sizeof(pt_frame_t));
+        cont->count = 0;
+
+        /* Prepare the `__gc` finalizer to free the stack. */
+        lua_newtable(L);
+        lua_pushcfunction(L, _free_resources);
+        lua_setfield(L, -2, "__gc");
+        lua_setmetatable(L, -2);
+
+        /* This is our finalizer which will reside in the value stack. */
+        lua_newtable(L);
+        lua_newtable(L);
+        lua_pushvalue(L, -3);
+
+        /* Our finalizer fn. */
+        lua_pushcclosure(L, pallene_tracer_finalizer, 1);
+        lua_setfield(L, -2, "__close");
+        lua_setmetatable(L, -2);
+
+        /* Se finalizer object to registry. */
+        lua_setfield(L, LUA_REGISTRYINDEX, PALLENE_TRACER_FINALIZER_ENTRY);
+
+        /* Set stack container to registry .*/
+        lua_setfield(L, LUA_REGISTRYINDEX, PALLENE_TRACER_CONTAINER_ENTRY);
+
+        /* The debug traceback fn. */
+        lua_register(L, "pallene_tracer_debug_traceback", pallene_tracer_debug_traceback);
+
+        /* Push the finalizer object in the stack. */
+        lua_getfield(L, LUA_REGISTRYINDEX, PALLENE_TRACER_FINALIZER_ENTRY);
+    } else {
+        cont = lua_touserdata(L, -1);
+        lua_getfield(L, LUA_REGISTRYINDEX, PALLENE_TRACER_FINALIZER_ENTRY);
+    }
+
+    return cont;
+}
+
+/* When Pallene call-stack is overflowed. */
+l_noret pallene_tracer_runtime_callstack_overflow_error(lua_State *L) {
+    luaL_error(L, "pallene callstack overflow");
+    PALLENE_TRACER_UNREACHABLE;
+}
```

Comment:
Does it make sense for this to be a recoverable error (can be caught with pcall?). If yes, there should be tests to ensure that it does something sensible on overflow. If not, we should probably do something else:
1. panic/abort
2. silently do not record stack frames after the limit. (Maybe record just the stack frame count)

Reply:
I will be adding tests as soon as I am done figure out a good, non-sucky stack number for Pallene tracer call-stack. It's too unstable now.

Reply:
Done.